ksqlDB should set key.converter to string for sink connectors if not specified by user

ksqlDB currently writes keys as strings, and as such any connector will fail if the worker is not set to use String for key converter. 

e.g.

```
ksql> DESCRIBE CONNECTOR SINK_POSTGRES_01;

Name                 : SINK_POSTGRES_01
Class                : io.confluent.connect.jdbc.JdbcSinkConnector
Type                 : sink
State                : RUNNING
WorkerId             : kafka-connect-01:8083

 Task ID | State  | Error Trace
---------------------------------------------------------------------------------------------------------------------------------------------------
 0       | FAILED | org.apache.kafka.connect.errors.ConnectException: Tolerance exceeded in error handler
        at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndHandleError(RetryWithToleranceOperator.java:178)
        at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execute(RetryWithToleranceOperator.java:104)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.convertAndTransformRecord(WorkerSinkTask.java:484)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.convertMessages(WorkerSinkTask.java:464)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:320)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:224)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:192)
        at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:177)
        at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:227)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.kafka.connect.errors.DataException: Failed to deserialize data for topic USERMENTIONS_COUNT to Avro:
        at io.confluent.connect.avro.AvroConverter.toConnectData(AvroConverter.java:110)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.lambda$convertAndTransformRecord$0(WorkerSinkTask.java:484)
        at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndRetry(RetryWithToleranceOperator.java:128)
        at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndHandleError(RetryWithToleranceOperator.java:162)
        ... 13 more
Caused by: org.apache.kafka.common.errors.SerializationException: Error deserializing Avro message for id -1
Caused by: org.apache.kafka.common.errors.SerializationException: Unknown magic byte!

---------------------------------------------------------------------------------------------------------------------------------------------------
```

This is a bad experience for the user because (a) it doesn't work and (b) it's hard to troubleshoot because the Kafka Connect error doesn't say it's the Key that's causing the deserialisation problem.

My suggestion: when the user issues `CREATE SINK CONNECTOR` ksqlDB should automagically set `'key.converter'       = 'org.apache.kafka.connect.storage.StringConverter' ` if it's not present in the user's DML.  As and when ksqlDB supports richer strings we can change this, but for now it will smooth the path _a lot_ for users if we can do this. There is precedent for similar config overrides in `CREATE SOURCE CONNECTOR` setting up Single Message Transforms.

--- a/ksql-engine/src/main/java/io/confluent/ksql/connect/supported/Connectors.java
+++ b/ksql-engine/src/main/java/io/confluent/ksql/connect/supported/Connectors.java
@@ -18,12 +18,14 @@ package io.confluent.ksql.connect.supported;
 import com.google.common.collect.ImmutableMap;
 import io.confluent.ksql.connect.Connector;
 import java.util.EnumSet;
+import java.util.HashMap;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.function.Function;
 import java.util.stream.Collectors;
 import org.apache.kafka.connect.runtime.rest.entities.ConnectorInfo;
+import org.apache.kafka.connect.storage.StringConverter;
 
 public enum Connectors implements SupportedConnector {
 
@@ -55,10 +57,16 @@ public enum Connectors implements SupportedConnector {
   }
 
   public static Map<String, String> resolve(final Map<String, String> configs) {
-    final SupportedConnector connector =
-        CONNECTORS.get(configs.get(CONNECTOR_CLASS));
-    return connector == null ? configs : connector.resolveConfigs(configs);
+    final SupportedConnector connector = CONNECTORS.get(configs.get(CONNECTOR_CLASS));
+    final Map<String, String> resolvedConfigs = new HashMap<>(
+        connector == null ? configs : connector.resolveConfigs(configs)
+    );
 
+    // at the moment, KSQL only supports String keys - when we
+    // support non-string keys we should remove this line
+    // (see: https://github.com/confluentinc/ksql/issues/3534)
+    resolvedConfigs.putIfAbsent("key.converter", StringConverter.class.getName());
+    return resolvedConfigs;
   }
 
   @Override

--- a/ksql-engine/src/main/java/io/confluent/ksql/connect/supported/JdbcSource.java
+++ b/ksql-engine/src/main/java/io/confluent/ksql/connect/supported/JdbcSource.java
@@ -24,7 +24,6 @@ import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import org.apache.kafka.connect.runtime.rest.entities.ConnectorInfo;
-import org.apache.kafka.connect.storage.StringConverter;
 import org.apache.kafka.connect.transforms.ExtractField;
 import org.apache.kafka.connect.transforms.ValueToKey;
 
@@ -69,7 +68,6 @@ public final class JdbcSource implements SupportedConnector {
       resolved.put("transforms.ksqlExtractString.field", key);
     }
 
-    resolved.putIfAbsent("key.converter", StringConverter.class.getName());
     resolved.putIfAbsent("tasks.max", "1");
     return resolved;
   }

--- a/ksql-engine/src/test/java/io/confluent/ksql/connect/supported/ConnectorsTest.java
+++ b/ksql-engine/src/test/java/io/confluent/ksql/connect/supported/ConnectorsTest.java
@@ -1,0 +1,38 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.connect.supported;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.hasEntry;
+
+import com.google.common.collect.ImmutableMap;
+import java.util.Map;
+import org.apache.kafka.connect.storage.StringConverter;
+import org.junit.Test;
+
+public class ConnectorsTest {
+
+  @Test
+  public void shouldResolveConfigs() {
+    // Given:
+    final Map<String, String> configs = ImmutableMap.of();
+
+    // When/Then:
+    assertThat(
+        Connectors.resolve(configs), hasEntry("key.converter", StringConverter.class.getName()));
+  }
+
+}

--- a/ksql-engine/src/test/java/io/confluent/ksql/connect/supported/JdbcSourceTest.java
+++ b/ksql-engine/src/test/java/io/confluent/ksql/connect/supported/JdbcSourceTest.java
@@ -133,7 +133,6 @@ public class JdbcSourceTest {
             .put("transforms.ksqlCreateKey.fields", "id")
             .put("transforms.ksqlExtractString.type", "org.apache.kafka.connect.transforms.ExtractField$Key")
             .put("transforms.ksqlExtractString.field", "id")
-            .put("key.converter", "org.apache.kafka.connect.storage.StringConverter")
             .put("tasks.max", "1")
             .build()));
   }

--- a/ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/execution/ConnectExecutorTest.java
+++ b/ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/execution/ConnectExecutorTest.java
@@ -16,11 +16,15 @@
 package io.confluent.ksql.rest.server.execution;
 
 import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.hasEntry;
 import static org.hamcrest.Matchers.instanceOf;
 import static org.mockito.ArgumentMatchers.anyMap;
 import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
+import static org.mockito.hamcrest.MockitoHamcrest.argThat;
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
@@ -36,6 +40,7 @@ import io.confluent.ksql.services.ConnectClient.ConnectResponse;
 import io.confluent.ksql.services.ServiceContext;
 import io.confluent.ksql.statement.ConfiguredStatement;
 import io.confluent.ksql.util.KsqlConfig;
+import java.util.Map;
 import java.util.Optional;
 import org.apache.http.HttpStatus;
 import org.apache.kafka.connect.runtime.rest.entities.ConnectorInfo;
@@ -43,6 +48,7 @@ import org.apache.kafka.connect.runtime.rest.entities.ConnectorType;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
+import org.mockito.ArgumentMatchers;
 import org.mockito.Mock;
 import org.mockito.junit.MockitoJUnitRunner;
 
@@ -72,6 +78,7 @@ public class ConnectExecutorTest {
     when(serviceContext.getConnectClient()).thenReturn(connectClient);
   }
 
+  @SuppressWarnings("unchecked")
   @Test
   public void shouldPassInCorrectArgsToConnectClient() {
     // Given:
@@ -81,7 +88,7 @@ public class ConnectExecutorTest {
     ConnectExecutor.execute(CREATE_CONNECTOR_CONFIGURED, ImmutableMap.of(), null, serviceContext);
 
     // Then:
-    verify(connectClient).create("foo", ImmutableMap.of("foo", "bar"));
+    verify(connectClient).create(eq("foo"), (Map<String, String>) argThat(hasEntry("foo", "bar")));
   }
 
   @Test

